Source/Core/Common/LdrWatcher.cpp

```diff
@@ -0,0 +1,178 @@
+                                    PCLDR_DLL_NOTIFICATION_DATA NotificationData, PVOID Context)
+{
+  if (NotificationReason != LDR_DLL_NOTIFICATION_REASON_LOADED)
+    return;
+  auto& data = NotificationData->Loaded;
+  auto observer = static_cast<const LdrObserver*>(Context);
+  LdrObserverRun(*observer, data.BaseDllName, reinterpret_cast<uintptr_t>(data.DllBase));
+}
+
+// This only works on Vista+. On lower platforms, it will be a no-op.
+class LdrDllNotifier
+{
+public:
+  static LdrDllNotifier& GetInstance()
+  {
+    static LdrDllNotifier notifier;
+    return notifier;
+  };
+  void Install(LdrObserver* observer);
+  void Uninstall(LdrObserver* observer);
+
+private:
+  LdrDllNotifier();
+  bool Init();
+  LdrRegisterDllNotification_t LdrRegisterDllNotification{};
+  LdrUnregisterDllNotification_t LdrUnregisterDllNotification{};
+  bool initialized{};
+};
+
+LdrDllNotifier::LdrDllNotifier()
+{
+  initialized = Init();
+}
+
+bool LdrDllNotifier::Init()
+{
+  auto ntdll = GetModuleHandleW(L"ntdll");
+  if (!ntdll)
+    return false;
+  LdrRegisterDllNotification = reinterpret_cast<decltype(LdrRegisterDllNotification)>(
+      GetProcAddress(ntdll, "LdrRegisterDllNotification"));
+  if (!LdrRegisterDllNotification)
+    return false;
+  LdrUnregisterDllNotification = reinterpret_cast<decltype(LdrUnregisterDllNotification)>(
+      GetProcAddress(ntdll, "LdrUnregisterDllNotification"));
+  if (!LdrUnregisterDllNotification)
+    return false;
+  return true;
+}
+
+void LdrDllNotifier::Install(LdrObserver* observer)
+{
+  if (!initialized)
+    return;
+  void* cookie{};
+  if (!NT_SUCCESS(LdrRegisterDllNotification(0, DllNotificationCallback,
+                                             static_cast<PVOID>(observer), &cookie)))
+    cookie = {};
+  observer->cookie = cookie;
+  return;
+}
+
+void LdrDllNotifier::Uninstall(LdrObserver* observer)
+{
+  if (!initialized)
+    return;
+  LdrUnregisterDllNotification(observer->cookie);
+  observer->cookie = {};
+  return;
+}
+
+LdrWatcher::~LdrWatcher()
+{
+  UninstallAll();
+}
+
+// Needed for RtlInitUnicodeString
+#pragma comment(lib, "ntdll")
+
+bool LdrWatcher::InjectCurrentModules(const LdrObserver& observer)
+{
+  // Use TlHelp32 instead of psapi functions to reduce dolphin's dependency on psapi
+  // (revisit this when Win7 support is dropped).
+  HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPMODULE, GetCurrentProcessId());
+  if (snapshot == INVALID_HANDLE_VALUE)
+    return false;
+  MODULEENTRY32 entry;
+  entry.dwSize = sizeof(entry);
+  for (BOOL rv = Module32First(snapshot, &entry); rv == TRUE; rv = Module32Next(snapshot, &entry))
+  {
+    UNICODE_STRING module_name;
+    RtlInitUnicodeString(&module_name, entry.szModule);
+    LdrObserverRun(observer, &module_name, reinterpret_cast<uintptr_t>(entry.modBaseAddr));
+  }
+  CloseHandle(snapshot);
+  return true;
+}
+
+void LdrWatcher::Install(const LdrObserver& observer)
+{
+  observers.emplace_back(std::move(observer));
```

Comment:
This would still result in a copy. move ctors/assignment operators take a `T&&`, but this would be a `const T&&`